bin/internal/shared.sh

```diff
@@ -140,6 +149,18 @@ function upgrade_flutter () (
     touch "$FLUTTER_ROOT/bin/cache/.dartignore"
     "$FLUTTER_ROOT/bin/internal/update_dart_sdk.sh"
 
+    local stamp="$(cat "$STAMP_PATH")"
+    local old_revision="${stamp%%:*}" old_flutter_tool_args="${stamp#*:}"
+    if [[ "$old_flutter_tool_args" == "$FLUTTER_TOOL_ARGS" ]] \
+        && git -C "$FLUTTER_ROOT" diff --quiet "$old_revision" "$revision" \
+             -- "${flutter_tool_source_pathspecs[@]}" >/dev/null 2>&1; then
+      # The cache has the same FLUTTER_TOOL_ARGS as we would compile,
+      # and a revision with the same tool source as we would compile.
+      # So that's a cache hit; just update the stamp.
+      echo "$compilekey" > "$STAMP_PATH"
+      exit $?
```

Comment:
Can you explain this bash wizardry? I thought that calling `exit` from within a function would exit the entire process. How are we reaching the end of this script to invoke the flutter tool snapshot?

Reply:
Yeah. That wizardry was here already — I wrote that line to match the other early exit a bit above it:
```bash
    # A different shell process might have updated the tool/SDK.
    if [[ -f "$SNAPSHOT_PATH" && -s "$STAMP_PATH" && "$(cat "$STAMP_PATH")" == "$compilekey" && "$FLUTTER_TOOLS_DIR/pubspec.yaml" -ot "$FLUTTER_TOOLS_DIR/pubspec.lock" ]]; then
      exit $?
    fi
```
but I think I can explain it.

The comments at the beginning and end of the function perhaps help:
```bash
# This function is always run in a subshell. Running the function in a subshell
# is required to make sure any lock directory is cleaned up by the exit trap in
# _wait_for_lock.
function upgrade_flutter () (

  # …

  # The exit here is extraneous since the function is run in a subshell, but
  # this serves as documentation that running the function in a subshell is
  # required to make sure any lock directory created by mkdir is cleaned up.
  exit $?
)
```

Then the very subtle bit of shell syntax that this function is relying on is: instead of enclosing the function's body in braces `{ … }`, it's enclosed in parens `( … )`. That causes the body to be executed in a subshell:
https://www.gnu.org/software/bash/manual/bash.html#Command-Grouping
which is the shell's term for forking a child process.

So then the `exit` exits the current process, as usual — and that means the subshell which is the function body.

Reply:
ahh, good catch

Reply:
And here's the docs for the syntax of a shell function definition itself:
https://www.gnu.org/software/bash/manual/bash.html#Shell-Functions

> The _body_ of the function is the compound command _compound-command_ (see [Compound Commands](https://www.gnu.org/software/bash/manual/bash.html#Compound-Commands)). That command is usually a _list_ enclosed between `{` and `}`, but may be any compound command listed above.

So in addition to enclosing in parens or braces, you could write a shell function's body as just an `if` command or `for` command or some others. (But probably shouldn't. For that matter, this function would probably be a lot clearer if it used braces too, and made the parens a lot more conspicuous:
```bash
function upgrade_flutter () {
  # This function's body runs in a subshell, as part of making sure
  # any lock directory is cleaned up by the exit trap in _wait_for_lock.
  (
    mkdir -p "$FLUTTER_ROOT/bin/cache"
    # … etc., etc., rest of this function's body …
  )
}
```
That'd behave the same as the existing code, but it'd be easier to see that something unusual is going on.)

Reply:
So when we `exit $?`, is the code we're exiting with from `echo "$compilekey" > "$STAMP_PATH"`?

Reply:
Right. Which is always 0, unless "$STAMP_PATH" was unwriteable or something.

(Also there's `set -e` at the top of the script, and I'm like 75% confident that that gets propagated to a subshell. In that case, if we reach the "exit" statement at all, "$?" has to be 0, and the same is true for the other exit statements in the function.)